File: Exercises/ExercisesFortran/08Jan_Chapter06/SingleChain/Source/grow.f
      Subroutine Grow(LOldChain,Weight,Ubonded,Unonb,Xst,Yst,Zst)
      Implicit None
Cccccccccccccccccccccccccccccccccccccccccccccccccccc
C     Grow A Chain Using Cbmc Or Random Insertion  C
Cccccccccccccccccccccccccccccccccccccccccccccccccccc

      Include 'system.inc'

      Logical LOldChain,LValidTrial
      Integer I,J,K,Ntrial,Iwalk

      Double Precision Ubonded,Unonb,Weight,Xst(Maxchain),
     &     Yst(Maxchain),Zst(Maxchain),Xtrial(Maxtrial),
     &     Ytrial(Maxtrial),Ztrial(Maxtrial),Dx1,Dx2,Dy1,
     &     Dy2,Dz1,Dz2,Ubend(Maxtrial),Uext(Maxtrial),X,Y,
     &     Z,Ws,RandomNumber,Cumw,Cmmm,U,Bfac(Maxtrial)

      Ubonded = 0.0d0
      Unonb   = 0.0d0
      Weight  = 1.0d0
      Ws      = 0.0d0

      If(Lcbmc) Then
         Ntrial = NumberOfTrialPositions
      Else
         Ntrial = 1
      Endif

C     the first bead of the trial chain is always at (0,0,0)
      Xst(1) = 0.0d0
      Yst(1) = 0.0d0
      Zst(1) = 0.0d0

Ccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccc
C     Grow The Chain; Loop Over All Segments Exept The First One  C
C                                                                 C
C     Xpos/Ypos/Zpos       : Position Of The Old Chain            C
C     Xst/Yst/Zst          : Position Of The Trial Chain          C
C     Xtrial/Ytrial/Ztrial : Position Of A Trial Segment          C
C     Lcbmc                : Do We Use Cbmc (.True. Or .False.)   C
C     LOldChain                 : Do We Have The Old Configuration     C
C                            (Only For Dynamic Schemes)           C
Ccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccc

C     Loop over all the particles in the polymer chain 
      Do I=2,ChainLength

C     Loop Over All Trial Positions   C
         Do J=1,Ntrial

 10         Ubend(J) = 0.0d0
            Uext(J)  = 0.0d0
            LValidTrial   = .False.
            Ws       = 0.0d0
            
Cccccccccccccccccccccccccccccccccccccccccccccccccccccccccccc
C     Generate Trial Position                              C
C     J=1 And LOldChain : Old Position                          C
C     Otherwise    : New Position                          C
C                                                          C
C     Old Coordinates Are Stored In Xpos/Ypos/Zpos         C
C     The New Chain Is Stored In Xst/Yst/Zst               C
C     Trial Positions Are Stored In Xtrial/Ytrial/Ztrial   C
C                                                          C
C     Subroutine Ran_Sphere(X,Y,Z): Creates A Random Unit  C
C     Vector On A Sphere centered at (0,0,0)               C
Cccccccccccccccccccccccccccccccccccccccccccccccccccccccccccc

C    if an old chain exists, we must retrace it to obtain its statistical weight; in that case the position of bead I in the old chain is also used as the first trial position of bead I

C    Note: retracing the old chain is not indispensable in our example because
c    we only have one chain; instead we could have just saved its statistical
C    weight from the previous step.
C    However, in a real CBMC implementation with more than one chain, the
C    environment of the chain may have changed since the previous time its
C    statistical weight was calculated; in that case, retracing is necessary. To
C    keep the algorithm general, CBMC is always implemented with retracing of the old
C    chain.
 
            If(J.Eq.1.And.LOldChain) Then
               Xtrial(J) = Xpos(I)
               Ytrial(J) = Ypos(I)
               Ztrial(J) = Zpos(I)
            Else  
C     Otherwise, generate a new trial position for bead I
C     Start Modification




C     End Modification
            Endif

Cccccccccccccccccccccccccccccccccccccccccc
C     Calculate Bond-Bending Potential   C
Cccccccccccccccccccccccccccccccccccccccccc

            If(I.Gt.2) Then
C              Vector from bead I-1 to trial position J
               Dx1 = Xtrial(J) - Xst(I-1)
               Dy1 = Ytrial(J) - Yst(I-1)
               Dz1 = Ztrial(J) - Zst(I-1)
                                         
C              Vector from bead I-1 to bead I-2 
               Dx2 = Xst(I-2)  - Xst(I-1)
               Dy2 = Yst(I-2)  - Yst(I-1)
               Dz2 = Zst(I-2)  - Zst(I-1)

C              Calculate angular potential energy  associated with 3 consecutive beads:  I-2 .....  I-1 ..... J
               Call Angle(Dx1,Dy1,Dz1,Dx2,Dy2,Dz2,U)

               Ubend(J) = U
            Endif

Cccccccccccccccccccccccccccccccccccccccccccccccccccccc
C     Accept Or Reject The Chosen Configuration      C
C     This Is For Cmbc Only, Because Only In Cbmc    C
C     A Trial Position According To A Bond-Bending   C
C     Potential Has To Be Generated...               C
C                                                    C
C     When J=1.And.LOldChain : Always Accepted !! (Why?)  C
Cccccccccccccccccccccccccccccccccccccccccccccccccccccc

            If(Lcbmc) Then

c              old chain is always accepted
               If(J.Eq.1.And.LOldChain) Then
                  LValidTrial = .True.
               Else
C              Start Modification
C              Accept or reject the other trial positions according to the Metropolis algorithm


 
C              End Modification
               Endif

C           If CMBC is not used, always accept  the single trial position that was randomly generated.
            Else
               LValidTrial = .True.
            Endif

            If(.Not.LValidTrial) Goto 10

Ccccccccccccccccccccccccccccccccccccccc
C     Calculate The External Energy   C
C     Only When There At Least 4      C
C     Beads...                        C
Ccccccccccccccccccccccccccccccccccccccc

            If(I.Gt.3) Then
               Do K=1,(I-3)

                  Dx1 = Xtrial(J) - Xst(K)
                  Dy1 = Ytrial(J) - Yst(K)
                  Dz1 = Ztrial(J) - Zst(K)
                  
                  Call Repulsion(Dx1,Dy1,Dz1,U)

                  Uext(J) = Uext(J) + U
               Enddo
            Endif

C        End Loop Over All Trial Positions           C
         Enddo

Ccccccccccccccccccccccccccccccccccccccccccccccccccc
C     Choose A Trial Position For Cbmc                 C
C     For Cbmc And An Old Config.: take 1st Config.    C
Ccccccccccccccccccccccccccccccccccccccccccccccccccc

         Iwalk = 1

         If(Lcbmc) Then
            Do K=1,Ntrial
               Bfac(K) = Dexp(-Beta*Uext(K))
               Ws      = Ws + Bfac(K)
            Enddo

            Cumw  = Bfac(1)
            Cmmm  = Ws*RandomNumber()

            If(.Not.LOldChain) Then
               Do While(Cumw.Lt.Cmmm)
                  Iwalk = Iwalk + 1
                  Cumw  = Cumw  + Bfac(Iwalk)
               Enddo
            Endif
         Else
            Ws = Dexp(-Beta*(Uext(1)+Ubend(1)))
         Endif

Ccccccccccccccccccccccccccccccccccccccccccc
C     Store The Chosen Configuration      C
C     Update Energies/Rosenbluth Weight   C
Ccccccccccccccccccccccccccccccccccccccccccc

         Xst(I) = Xtrial(Iwalk)
         Yst(I) = Ytrial(Iwalk)
         Zst(I) = Ztrial(Iwalk)

         Ubonded     = Ubonded + Ubend(Iwalk)
         Unonb       = Unonb   + Uext(Iwalk)
         Weight      = Weight*Ws/Dble(Ntrial)

C     END loop over all particles in chain
      Enddo

      Return
      End

File: Exercises/ExercisesC/08Jan_Chapter06/SingleChain/Source/grow.c
#include <stdlib.h>
#include <stdio.h>
#include <math.h>
#include "system.h"
#include "ran_uniform.h"

// Calculate angular potential energy  associated with 3 consecutive beads:  I-2 .....  I-1 ..... I
double Angle(VECTOR dr1,VECTOR dr2)
{
  double a,U;
  // dr1:          vector from I-1 to I
  // dr2:          vector from I-1 to I-2
  // U:            angular potential energy
  // a:            cosine of the angle formed by the three beads

  //      Calculate the cosine a of the angle formed by the  vectors dr1 and dr2 by calculating their dot product:
  //         dr1.dr2 = |dr1| |dr2| cos(dr1^dr2) <=>
  //     <=> cos(dr1^dr2) = dr1.dr2/(|dr1|*|dr2|)

  a=(dr1.x*dr2.x+dr1.y*dr2.y+dr1.z*dr2.z)/
      sqrt((SQR(dr1.x)+SQR(dr1.y)+SQR(dr1.z))*
           (SQR(dr2.x)+SQR(dr2.y)+SQR(dr2.z)));

  // Calculate the potential energy associated with that angle
  //    Note: the function acos(a) returns the angle between 0 and Pi associated
  //    with a cosine value of R
  U=0.5*Kt*SQR(acos(a)-Theta0);
  return U;
}

// Calculate repulsive potential energy between two beads
double Repulsion(VECTOR dr)
{
//       Dr:       vector connecting the two beads
//       U:        repulsive energy between the two beads
//       R:        distance between two beads
  double R,U;

  R=sqrt(SQR(dr.x)+SQR(dr.y)+SQR(dr.z));
  U=0.0;
  // If the distance between the two beads is less than cutoff, calculate the repulsive energy.
  //   Otherwise, the repulsive energy is 0.

  if(R<Rcut) U=A*SQR(R-Rcut)/SQR(Rcut);
  return U;
}

// grow a chain using cbmc or random insertion
void Grow(int LOldChain,double *Weight,double *Ubonded,double *Unonb)
{
  int LValidTrial;
  int i,j,k,Ntrial,Iwalk;
  VECTOR dr1,dr2,vec,Trial[MAX_TRIALS];
  double BendEnergy[MAX_TRIALS],ExternalEnergy[MAX_TRIALS],BoltzmannFactor[MAX_TRIALS];
  double Ws,Cumw,Cmmm;

  *Ubonded=0.0;
  *Unonb=0.0;
  *Weight=1.0;
  Ws=0.0;

  if(Lcbmc)
    Ntrial=NumberOfTrialPositions;
  else
    Ntrial=1;

  // the first bead of the trial chain is always at (0,0,0)
  TrialPositions[0].x=0.0;
  TrialPositions[0].y=0.0;
  TrialPositions[0].z=0.0;

  // Grow The Chain; Loop Over All Segments Exept The First One
  //
  // Positions            : Position Of The Old Chain
  // TrialPositions       : Position Of The Trial Chain
  // Trial                : Position Of A Trial Segment
  // Lcbmc                : Do We Use Cbmc (.True. Or .False.)
  // LOldChain            : Do We Have The Old Configuration
  //                        (Only For Dynamic Schemes)

  // Loop over all the particles in the polymer chain
  for(i=1;i<ChainLength;i++)
  {
    // loop over all trial positions
    for(j=0;j<Ntrial;j++)
    {
      do
      {
        BendEnergy[j]=0.0;
        ExternalEnergy[j]=0.0;
        LValidTrial=FALSE;

        // generate trial position
        // j=0 And LOldChain : old position
        // Otherwise    : new position
        // old coordinates are stored in Positions
        // the new chain is stored in TrialPositions
        // trial positions are stored in Trial
        // function RanSphere(&vec): creates a random unit
        // vector on a sphere centered at (0,0,0)

          // if an old chain exists, we must retrace it to obtain its
          // statistical weight; in that case the position of bead I in the old chain is also used as the first trial position of bead I

          // Note: retracing the old chain is not indispensable in our example
          // because we only have one chain; instead we could have just saved its statistical
          // weight from the previous step.
          // However, in a real CBMC implementation with more than one chain, the
          // environment of the chain may have changed since the previous time its
          // statistical weight was calculated; in that case, retracing is necessary. To
          // keep the algorithm general, CBMC is always implemented with retracing of
          // the old chain.

          if(j==0&&LOldChain)
          {
            Trial[j].x=Positions[i].x;
            Trial[j].y=Positions[i].y;
            Trial[j].z=Positions[i].z;
          }
          // Otherwise, generate a new trial position for bead I
          else
          {
        // start modification

            RanSphere(&vec);

            Trial[j].x=TrialPositions[i-1].x+vec.x;
            Trial[j].y=TrialPositions[i-1].y+vec.y;
            Trial[j].z=TrialPositions[i-1].z+vec.z;

        // end modification
          }

        // calculate bond-bending potential
        if(i>1)
        {
          // Vector from bead I-1 to trial position J
          dr1.x=Trial[j].x-TrialPositions[i-1].x;
          dr1.y=Trial[j].y-TrialPositions[i-1].y;
          dr1.z=Trial[j].z-TrialPositions[i-1].z;
          // Vector from bead I-1 to bead I-2
          dr2.x=TrialPositions[i-2].x-TrialPositions[i-1].x;
          dr2.y=TrialPositions[i-2].y-TrialPositions[i-1].y;
          dr2.z=TrialPositions[i-2].z-TrialPositions[i-1].z;
          // Calculate angular potential energy  associated with 3 consecutive beads:  I-2 .....  I-1 ..... J
          BendEnergy[j]=Angle(dr1,dr2);
        }

        // accept or reject the chosen configuration
        // this is for cmbc only, because only in cbmc
        // a trial position according to a bond-bending
        // potential has to be generated...
        // when j=0&&LOldChain : always accepted !! (why?)

        if(Lcbmc)
        {
            if(j==0&&LOldChain) // old chain is always accepted
              LValidTrial=TRUE;
            else
            {
          // start modification
          // Accept or reject the other trial positions according to the Metropolis algorithm
              if(RandomNumber()<exp(-Beta*(BendEnergy[j]))) {
                LValidTrial=TRUE;
              }

          // end modification
            }
        }
        else  // If CMBC is not used, always accept  the single trial position that was randomly generated.
          LValidTrial=TRUE;
      } while(!LValidTrial);

      // calculate the external energy
      // only when there are at least 4 beads...
      if(i>2)
      {
        for(k=0;k<(i-2);k++)
        {
          dr1.x=Trial[j].x-TrialPositions[k].x;
          dr1.y=Trial[j].y-TrialPositions[k].y;
          dr1.z=Trial[j].z-TrialPositions[k].z;
          ExternalEnergy[j]+=Repulsion(dr1);
        }
      }
    }

    // end loop over all trial positions

    // choose a trial position for cbmc
    // for cbmc and an old config.: 1st config.
    Ws=0.0;
    Iwalk=0;
    if(Lcbmc)
    {
      for(k=0;k<Ntrial;k++)
      {
        BoltzmannFactor[k]=exp(-Beta*ExternalEnergy[k]);
        Ws+=BoltzmannFactor[k];
      }
      Cumw=BoltzmannFactor[0];
      Cmmm=Ws*RandomNumber();
      if(!LOldChain)
      {
        while(Cumw<Cmmm)
        {
          Iwalk++;
          Cumw+=BoltzmannFactor[Iwalk];
        }
      }
    }
    else
      Ws=exp(-Beta*(ExternalEnergy[0]+BendEnergy[0]));

    // store the chosen configuration
    // update energies/Rosenbluth weight
    TrialPositions[i].x=Trial[Iwalk].x;
    TrialPositions[i].y=Trial[Iwalk].y;
    TrialPositions[i].z=Trial[Iwalk].z;

    (*Ubonded)+=BendEnergy[Iwalk];
    (*Unonb)+=ExternalEnergy[Iwalk];
    (*Weight)*=(Ws/Ntrial);
  } // END loop over all particles in chain
}
